File: python/day-04/camp_cleanup.py
import re


def ex1_and_ex2(text):
    items = re.compile(',|-|\n').split(text)

    ex1, ex2 = 0, 0

    for idx in range(0, len(items), 4):
        x1, x2 = int(items[idx]), int(items[idx + 1]) + 1
        y1, y2 = int(items[idx + 2]), int(items[idx + 3]) + 1
        range1, range2 = set(range(x1, x2)), set(range(y1, y2))

        if range1.issubset(range2) or range2.issubset(range1):
            ex1 += 1

        if set() != range1.intersection(range2):
            ex2 += 1

    return ex1, ex2


if __name__ == '__main__':
    ex_inp = open('./lib/day-04/ex-inp.txt').read()
    inp = open('./../inputs/day-04/inp.txt').read()

    print(ex1_and_ex2(ex_inp))  # (2, 4)
    print(ex1_and_ex2(inp))  # (595, 952)

File: elixir/lib/day-04/camp_cleanup.ex
defmodule CampCleanup do
  def ex1(text) do
    count_overlaps(text, fn range1, range2 ->
      MapSet.subset?(range1, range2) or MapSet.subset?(range2, range1)
    end)
  end

  def ex2(text) do
    count_overlaps(text, fn range1, range2 ->
      not Enum.empty?(MapSet.intersection(range1, range2))
    end)
  end

  defp count_overlaps(text, overlapped?) do
    text
    |> String.split(~r/,|-|\n/)
    |> Enum.map(&String.to_integer/1)
    |> Enum.chunk_every(4)
    |> Enum.count(fn [x1, x2, y1, y2] -> overlapped?.(MapSet.new(x1..x2), MapSet.new(y1..y2)) end)
  end
end

ex_inp = File.read!("./lib/day-04/ex-inp.txt")
inp = File.read!("./lib/day-04/inp.txt")

# 2
ex_inp |> CampCleanup.ex1() |> IO.puts()
# 595
inp |> CampCleanup.ex1() |> IO.puts()
# 4
ex_inp |> CampCleanup.ex2() |> IO.puts()
# 952
inp |> CampCleanup.ex2() |> IO.puts()
